File: idscp2/src/main/java/de/fhg/aisec/ids/idscp2/drivers/default_driver_impl/secure_channel/server/TLSServerThread.java
package de.fhg.aisec.ids.idscp2.drivers.default_driver_impl.secure_channel.server;

import de.fhg.aisec.ids.idscp2.drivers.default_driver_impl.secure_channel.TLSSessionVerificationHelper;
import de.fhg.aisec.ids.idscp2.idscp_core.FastLatch;
import de.fhg.aisec.ids.idscp2.idscp_core.configuration.SecureChannelInitListener;
import de.fhg.aisec.ids.idscp2.idscp_core.idscp_server.ServerConnectionListener;
import de.fhg.aisec.ids.idscp2.idscp_core.secure_channel.SecureChannel;
import de.fhg.aisec.ids.idscp2.idscp_core.secure_channel.SecureChannelEndpoint;
import de.fhg.aisec.ids.idscp2.idscp_core.secure_channel.SecureChannelListener;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.net.ssl.HandshakeCompletedEvent;
import javax.net.ssl.HandshakeCompletedListener;
import javax.net.ssl.SSLPeerUnverifiedException;
import javax.net.ssl.SSLSocket;
import java.io.*;
import java.net.SocketTimeoutException;
import java.util.concurrent.CompletableFuture;

/**
 * A TLSServerThread that notifies an IDSCP2Config when a secure channel was created and the
 * TLS handshake is done
 * <p>
 * When new data are available the serverThread transfers it to the SecureChannelListener
 *
 * @author Leon Beckmann (leon.beckmann@aisec.fraunhofer.de)
 */
public class TLSServerThread extends Thread implements HandshakeCompletedListener, SecureChannelEndpoint, Closeable {
    private static final Logger LOG = LoggerFactory.getLogger(TLSServerThread.class);

    private volatile boolean running = true;
    private final DataInputStream in;
    private final DataOutputStream out;
    private final SSLSocket sslSocket;
    private final SecureChannelInitListener configCallback;
    private final CompletableFuture<ServerConnectionListener> serverListenerPromise;
    private final CompletableFuture<SecureChannelListener> channelListenerPromise = new CompletableFuture<>();
    private final FastLatch tlsVerificationLatch = new FastLatch();

    TLSServerThread(SSLSocket sslSocket, SecureChannelInitListener configCallback,
                    CompletableFuture<ServerConnectionListener> serverListenerPromise) throws IOException {
        this.sslSocket = sslSocket;
        this.configCallback = configCallback;
        this.serverListenerPromise = serverListenerPromise;
        // Set timeout for blocking read
        sslSocket.setSoTimeout(5000);
        in = new DataInputStream(sslSocket.getInputStream());
        out = new DataOutputStream(sslSocket.getOutputStream());
    }

    @Override
    public void run() {
        // first run the tls handshake to enforce catching every error occurred during the handshake
        // before reading from buffer. Else if there exists any non-catched exception during handshake
        // the thread would wait forever in onError() until handshakeCompleteListener is called
        try {
            sslSocket.startHandshake();
            // Wait for TLS session verification
            tlsVerificationLatch.await();
        } catch (IOException e) {
            LOG.warn("Exception occurred during SSL handshake. Quiting server thread...", e);
            running = false;
        }

        //wait for new data while running
        byte[] buf;
        while (running) {
            try {
                int len = in.readInt();
                buf = new byte[len];
                in.readFully(buf, 0, len);
                onMessage(buf);
            } catch (SocketTimeoutException ignore) {
                // Timeout catches safeStop() call and allows to send server_goodbye
            } catch (EOFException e) {
                onClose();
                running = false;
            } catch (IOException e) {
                onError();
                running = false;
            }
        }
        closeSockets();
    }

    private void closeSockets() {
        try {
            out.close();
            in.close();
            sslSocket.close();
        } catch (IOException ignore) {
        }
    }

    @Override
    public boolean send(byte[] data) {
        if (!isConnected()) {
            LOG.error("Server cannot send data because socket is not connected");
            closeSockets();
            return false;
        } else {
            try {
                out.writeInt(data.length);
                out.write(data);
                out.flush();
                LOG.trace("Send message: " + new String(data));
                return true;
            } catch (IOException e) {
                LOG.error("ServerThread cannot send data.");
                closeSockets();
                return false;
            }
        }
    }

    private void onClose() {
        channelListenerPromise.thenAccept(SecureChannelListener::onClose);
    }

    private void onError() {
        channelListenerPromise.thenAccept(SecureChannelListener::onError);
    }

    @Override
    public void close() {
        safeStop();
    }

    @Override
    public void onMessage(byte[] data) {
        channelListenerPromise.thenAccept(listener -> listener.onMessage(data));
    }

    private void safeStop() {
        running = false;
    }

    public boolean isConnected() {
        return (sslSocket != null && sslSocket.isConnected());
    }

    @Override
    public void handshakeCompleted(HandshakeCompletedEvent handshakeCompletedEvent) {
        if (LOG.isDebugEnabled()) {
            LOG.debug("TLS Handshake was successful");
        }

        // verify tls session on application layer: hostname verification, certificate validity
        try {
            TLSSessionVerificationHelper.verifyTlsSession(handshakeCompletedEvent.getSession());
            LOG.debug("TLS session is valid");
        } catch (SSLPeerUnverifiedException e) {
            if (LOG.isWarnEnabled()) {
                LOG.warn("TLS session is not valid. Close TLS connection", e);
            }
            running = false;  // set running false before tlsVerificationLatch is decremented
            return;
        } finally {
            tlsVerificationLatch.unlock();
        }

        //provide secure channel to IDSCP2 Config and register secure channel as listener
        SecureChannel secureChannel = new SecureChannel(this);
        this.channelListenerPromise.complete(secureChannel);
        configCallback.onSecureChannel(secureChannel, serverListenerPromise);
    }
}

File: idscp2/src/main/kotlin/de/fhg/aisec/ids/idscp2/default_drivers/secure_channel/tlsv1_3/server/TLSServerThread.kt
package de.fhg.aisec.ids.idscp2.default_drivers.secure_channel.tlsv1_3.server

import de.fhg.aisec.ids.idscp2.default_drivers.secure_channel.tlsv1_3.TLSSessionVerificationHelper
import de.fhg.aisec.ids.idscp2.idscp_core.FastLatch
import de.fhg.aisec.ids.idscp2.idscp_core.api.idscp_connection.Idscp2Connection
import de.fhg.aisec.ids.idscp2.idscp_core.api.idscp_server.SecureChannelInitListener
import de.fhg.aisec.ids.idscp2.idscp_core.secure_channel.SecureChannel
import de.fhg.aisec.ids.idscp2.idscp_core.drivers.SecureChannelEndpoint
import de.fhg.aisec.ids.idscp2.idscp_core.secure_channel.SecureChannelListener
import de.fhg.aisec.ids.idscp2.idscp_core.api.idscp_server.ServerConnectionListener
import org.slf4j.LoggerFactory
import java.io.*
import java.net.SocketTimeoutException
import java.util.concurrent.CompletableFuture
import javax.net.ssl.HandshakeCompletedEvent
import javax.net.ssl.HandshakeCompletedListener
import javax.net.ssl.SSLPeerUnverifiedException
import javax.net.ssl.SSLSocket

/**
 * A TLSServerThread that notifies an IDSCP2Config when a secure channel was created and the
 * TLS handshake is done
 *
 *
 * When new data are available the serverThread transfers it to the SecureChannelListener
 *
 * @author Leon Beckmann (leon.beckmann@aisec.fraunhofer.de)
 */
class TLSServerThread<CC : Idscp2Connection> internal constructor(
        private val sslSocket: SSLSocket,
        private val configCallback: SecureChannelInitListener<CC>,
        private val serverListenerPromise: CompletableFuture<ServerConnectionListener<CC>>) :
        Thread(), HandshakeCompletedListener, SecureChannelEndpoint, Closeable
{
    @Volatile
    private var running = true
    private val `in`: DataInputStream
    private val out: DataOutputStream
    private val channelListenerPromise = CompletableFuture<SecureChannelListener>()
    private val tlsVerificationLatch = FastLatch()
    override fun run() {
        // first run the tls handshake to enforce catching every error occurred during the handshake
        // before reading from buffer. Else if there exists any non-catched exception during handshake
        // the thread would wait forever in onError() until handshakeCompleteListener is called
        try {
            sslSocket.startHandshake()
            // Wait for TLS session verification
            tlsVerificationLatch.await()
        } catch (e: Exception) {
            LOG.warn("Exception occurred during SSL handshake. Quiting server thread...", e)
            onError(e)
            running = false
        }

        //wait for new data while running
        var buf: ByteArray
        while (running) {
            try {
                val len = `in`.readInt()
                buf = ByteArray(len)
                `in`.readFully(buf, 0, len)
                onMessage(buf)
            } catch (ignore: SocketTimeoutException) {
                // Timeout catches safeStop() call and allows to send server_goodbye
            } catch (e: EOFException) {
                onClose()
                running = false
            } catch (e: IOException) {
                onError(e)
                running = false
            }
        }
        closeSockets()
    }

    private fun closeSockets() {
        try {
            out.close()
            `in`.close()
            sslSocket.close()
        } catch (ignore: IOException) {
        }
    }

    override fun send(bytes: ByteArray): Boolean {
        return if (!isConnected) {
            LOG.warn("Server cannot send data because socket is not connected")
            closeSockets()
            false
        } else {
            try {
                out.writeInt(bytes.size)
                out.write(bytes)
                out.flush()
                true
            } catch (e: IOException) {
                LOG.warn("Server could not send data", e)
                closeSockets()
                false
            }
        }
    }

    private fun onClose() {
        channelListenerPromise.thenAccept { obj: SecureChannelListener -> obj.onClose() }
    }

    private fun onError(t: Throwable) {
        channelListenerPromise.thenAccept { secureChannelListener: SecureChannelListener -> secureChannelListener.onError(t) }
    }

    override fun close() {
        safeStop()
    }

    fun onMessage(bytes: ByteArray) {
        channelListenerPromise.thenAccept { listener: SecureChannelListener -> listener.onMessage(bytes) }
    }

    private fun safeStop() {
        running = false
    }

    override val isConnected: Boolean
        get() = sslSocket.isConnected

    override fun handshakeCompleted(handshakeCompletedEvent: HandshakeCompletedEvent) {
        if (LOG.isTraceEnabled) {
            LOG.trace("TLS Handshake was successful")
        }

        // verify tls session on application layer: hostname verification, certificate validity
        try {
            TLSSessionVerificationHelper.verifyTlsSession(handshakeCompletedEvent.session)
            if (LOG.isTraceEnabled) {
                LOG.trace("TLS session is valid")
            }
        } catch (e: SSLPeerUnverifiedException) {
            LOG.warn("TLS session is not valid. Close TLS connection", e)
            running = false // set running false before tlsVerificationLatch is decremented
            return
        } finally {
            tlsVerificationLatch.unlock()
        }

        //provide secure channel to IDSCP2 Config and register secure channel as listener
        val secureChannel = SecureChannel(this)
        channelListenerPromise.complete(secureChannel)
        configCallback.onSecureChannel(secureChannel, serverListenerPromise)
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(TLSServerThread::class.java)
    }

    init {
        // Set timeout for blocking read
        sslSocket.soTimeout = 5000
        `in` = DataInputStream(sslSocket.inputStream)
        out = DataOutputStream(sslSocket.outputStream)
    }
}